src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -1079,18 +1087,97 @@ public String Replace(char oldChar, char newChar)
             }
         }
 
-        // This method contains the same functionality as StringBuilder Replace. The only difference is that
-        // a new String has to be allocated since Strings are immutable
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        private extern String ReplaceInternal(String oldValue, String newValue);
-
-        public String Replace(String oldValue, String newValue)
+        public string Replace(string oldValue, string newValue)
         {
             if (oldValue == null)
                 throw new ArgumentNullException(nameof(oldValue));
-            // Note that if newValue is null, we treat it like String.Empty.
+            if (oldValue.Length == 0)
+                throw new ArgumentException(SR.Format(SR.Argument_StringZeroLength, nameof(oldValue)));
+
+            // Api behavior: if newValue is null, instances of oldValue are to be removed.
+            if (newValue == null)
+                newValue = string.Empty;
+
+            // Workaround for https://github.com/dotnet/coreclr/issues/16197
+            // Span<int> initialSpan = stackalloc int[StackallocIntBufferSizeLimit];
```

Comment:
Have we considered having a constructor on ValueListBuilder that takes care of stackallocing some initial span possibly with a hint as to the size? That would save callers repeating that work.

Reply:
You cannot do that unfortunately. The stackallocated memory is gone when the method returns. So if you stackalloc inside ValueListBuilder constructor, the memory is gone when the constructor returns.

Reply:
Ah of course.